src/scheduler/gc_works.rs

```diff
@@ -0,0 +1,462 @@
+        } else {
+            mmtk.scheduler
+                .add_coordinator_work(StopMutators::<E>::new(), worker);
+        }
+    }
+}
+
+impl<E: ProcessEdgesWork> CoordinatorWork<MMTK<E::VM>> for StopMutators<E> {}
+
+#[derive(Default)]
+pub struct EndOfGC;
+
+impl<VM: VMBinding> GCWork<VM> for EndOfGC {
+    fn do_work(&mut self, worker: &mut GCWorker<VM>, mmtk: &'static MMTK<VM>) {
+        mmtk.plan.common().base.set_gc_status(GcStatus::NotInGC);
+        <VM as VMBinding>::VMCollection::resume_mutators(worker.tls);
+    }
+}
+
+impl<VM: VMBinding> CoordinatorWork<MMTK<VM>> for EndOfGC {}
+
+static SCANNED_STACKS: AtomicUsize = AtomicUsize::new(0);
+
+#[derive(Default)]
+pub struct ScanStackRoots<Edges: ProcessEdgesWork>(PhantomData<Edges>);
+
+impl<E: ProcessEdgesWork> ScanStackRoots<E> {
+    pub fn new() -> Self {
+        Self(PhantomData)
+    }
+}
+
+impl<E: ProcessEdgesWork> GCWork<E::VM> for ScanStackRoots<E> {
+    fn do_work(&mut self, worker: &mut GCWorker<E::VM>, mmtk: &'static MMTK<E::VM>) {
+        trace!("ScanStackRoots");
+        <E::VM as VMBinding>::VMScanning::scan_thread_roots::<E>();
+        <E::VM as VMBinding>::VMScanning::notify_initial_thread_scan_complete(false, worker.tls);
+        mmtk.plan.common().base.set_gc_status(GcStatus::GcProper);
+    }
+}
+
+pub struct ScanStackRoot<Edges: ProcessEdgesWork>(
+    pub &'static mut Mutator<SelectedPlan<Edges::VM>>,
+);
+
+impl<E: ProcessEdgesWork> GCWork<E::VM> for ScanStackRoot<E> {
+    fn do_work(&mut self, worker: &mut GCWorker<E::VM>, mmtk: &'static MMTK<E::VM>) {
+        trace!("ScanStackRoot for mutator {:?}", self.0.get_tls());
+        <E::VM as VMBinding>::VMScanning::scan_thread_root::<E>(
+            unsafe { &mut *(self.0 as *mut _) },
+            worker.tls,
+        );
+        self.0.flush();
+        let old = SCANNED_STACKS.fetch_add(1, Ordering::SeqCst);
+        if old + 1 == <E::VM as VMBinding>::VMActivePlan::number_of_mutators() {
+            SCANNED_STACKS.store(0, Ordering::SeqCst);
+            <E::VM as VMBinding>::VMScanning::notify_initial_thread_scan_complete(
+                false, worker.tls,
+            );
+            mmtk.plan.common().base.set_gc_status(GcStatus::GcProper);
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct ScanVMSpecificRoots<Edges: ProcessEdgesWork>(PhantomData<Edges>);
+
+impl<E: ProcessEdgesWork> ScanVMSpecificRoots<E> {
+    pub fn new() -> Self {
+        Self(PhantomData)
+    }
+}
+
+impl<E: ProcessEdgesWork> GCWork<E::VM> for ScanVMSpecificRoots<E> {
+    fn do_work(&mut self, _worker: &mut GCWorker<E::VM>, _mmtk: &'static MMTK<E::VM>) {
+        trace!("ScanStaticRoots");
+        <E::VM as VMBinding>::VMScanning::scan_vm_specific_roots::<E>();
+    }
+}
+
+pub struct ProcessEdgesBase<E: ProcessEdgesWork> {
+    pub edges: Vec<Address>,
+    pub nodes: Vec<ObjectReference>,
+    pub mmtk: Option<&'static MMTK<E::VM>>,
+    worker: *mut GCWorker<E::VM>,
```

Comment:
This is for fast pointer dereferencing. We need to dereference this pointer once for each object copy (to get work-local allocators), which is highly frequent for semispace.

We can't use `&'static mut GCWorker` because we don't know which one is the execution worker when this work-packet is created. But using `Option<&'static mut GCWorker>` will introduce one branch, which may down the performance if we have frequent pointer dereferencing (or probably not too much effect?).

Probably I can use `MaybeUninit` or `UnsafeCell` instead of using raw pointers. Or just use `Option<&'static mut GCWorker>`.

Reply:
Probably just add the above as comments so we know the reason behind this pointer use.

Reply:
Done.